BUG: fixing circular import in daal4py/sklearnex device_offloading

samir-nasibli opened the pull request:
### Description 
Fixing circular import in daal4py/sklearnex device offloading.
The best way to resolve the issue here is to refactor the code to avoid it. This PR rethinking the design of sklearnex/daal4py/onedal4py import modules and their dependencies.

#### Changes proposed in the PR:
* adding _config for `onedal4py`, just for exposing some `sklearnex`'s config settings into `onedal4py` level
* remove circular import in `daal4py`/`onedal4py`/`sklearnex`:
  * removing _device_offloading module from `daal4py`, since after #1770 and adding #1843 there is no need for GPU offloading via daal4py syc_context.
  *  most of device_ofloading functionality moved to `onedal4py` level.
  * created _config module in onedal4py, for exposing some sklearnex config setting into onedal4py level, reused it on sklearnex level
  * `sklearnex` depends on `onedal4py` _config and _device_oflload modules.

![image](https://github.com/intel/scikit-learn-intelex/assets/19879134/e0447cd0-5e97-428c-b971-d96398eafd0c)

__Note:__ It is expected that it will not offload GPU validation Kmeans on main branch, since they are using daal4py GPU offloading, that is removed to call.
#1770 is considered fixe for this.

### TODO
* create sklearnex estimators, that uses daal4py backend
  * https://github.com/intel/scikit-learn-intelex/pull/1832#discussion_r1648578243

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes, if necessary (updated in # - _add PR number_)
- [x] The unit tests pass successfully.
- [X] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [x] I have added a respective label(s) to PR if I have a permission for that.

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
/intelci: run

samir-nasibli commented on daal4py/sklearn/__init__.py:
```diff
@@ -14,13 +14,16 @@
 from .monkeypatch.dispatcher import patch_is_enabled as sklearn_is_patched
 
 __all__ = [
+    "_get_config",
+    "_set_config",
```
will remove this. No need for public iface

samir-nasibli replied:
outdated

icfaust reviewed (commented):
In the case we drop in the daal4py oneapi interfaces, is this movement of device_offload to daal4py necessary? Couldn't we just delete the sklearnex import of various aspects in daal4py/sklearn/_device_offload.py and it would remove the circular import problem? If true, how far away are we from being able to do that.  As far as I see, none of the imported estimators from daal4py in sklearnex outside of Ridge can offload to GPU via daal4py. Its a question over aspects of:
Ridge
Lasso
ElasticNet
PairwiseDistances
train_test_split
roc_auc_score

or am I missing something?

In general just some minor naming convention stuff.  Would like the thoughts of other reviewers on that.

samir-nasibli commented on daal4py/sklearn/_device_offload.py:
```diff
@@ -14,26 +14,159 @@
+
+from ._config import _get_config
+
 try:
```
Actually it doesn't do replication. I have moved necessary primitives to daal4py level to avoid code duplication and import of sklearnex on daal4py side

samir-nasibli commented on daal4py/sklearn/_config.py:
```diff
@@ -0,0 +1,72 @@
+# limitations under the License.
+# ==============================================================================
+
+"""Tools to expose sklearnex's config settings to daal4py level."""
```
I think it this case for correct passing sklearnex config settings it should be threading safe here

samir-nasibli commented on daal4py/sklearn/_config.py:
```diff
@@ -0,0 +1,72 @@
+    return _threadlocal.d4p_global_config
+
+
+def _get_config():
```
As I have comment it above, I will remove it from __init__. This was not intended to be of public iface.
This will be only internally used. 
From my opinion naming is fine in case of consistency for sklearnex config. I probably will reduce some part of this file, but the rest is fine

samir-nasibli replied:
outdated

samir-nasibli commented:
> In the case we drop in the daal4py oneapi interfaces, is this movement of device_offload to daal4py necessary? Couldn't we just delete the sklearnex import of various aspects in daal4py/sklearn/_device_offload.py and it would remove the circular import problem? If true, how far away are we from being able to do that. As far as I see, none of the imported estimators from daal4py in sklearnex outside of Ridge can offload to GPU via daal4py. Its a question over aspects of: Ridge Lasso ElasticNet PairwiseDistances train_test_split roc_auc_score
> 
> or am I missing something?
> 
> In general just some minor naming convention stuff. Would like the thoughts of other reviewers on that.



> In the case we drop in the daal4py oneapi interfaces, is this movement of device_offload to daal4py necessary? Couldn't we just delete the sklearnex import of various aspects in daal4py/sklearn/_device_offload.py and it would remove the circular import problem? If true, how far away are we from being able to do that. As far as I see, none of the imported estimators from daal4py in sklearnex outside of Ridge can offload to GPU via daal4py. Its a question over aspects of: Ridge Lasso ElasticNet PairwiseDistances train_test_split roc_auc_score
> 
> or am I missing something?
> 
> In general just some minor naming convention stuff. Would like the thoughts of other reviewers on that.

@icfaust Thank you for the review!
I think there are several things mixed up here.
Perhaps this is a lack of of my PR description. I want to bring some clarity.

* This PR tries to keep everything as it is. It does not imply changing anything in flow device offloading.
* For oneapi iface we have onedal4py+sklearnex. If we drop daal4py, then all these necessary things will simply be transferred to the sklearnex _device_offload module, since daal4py will no longer be needed.

Some minor updates are required, but overall PR is ready

samir-nasibli commented:
> As far as I see, none of the imported estimators from daal4py in sklearnex outside of Ridge can offload to GPU via daal4py. Its a question over aspects of: Ridge Lasso ElasticNet PairwiseDistances train_test_split roc_auc_score
> 
> or am I missing something?

If I understood the question correctly, and if you asking why we need daal4py sycl_context, so then actually Kmeans and other estimators from daal4py.sklearn that is used on sklearnex main branch does GPU offloading via sycl_context. so that is why we need to leave it as is. 

If anything requires drop I would like to do this reduction in the next steps.

samir-nasibli commented:
Dear reviewers.
In my PR, I tried to get rid of circular imports. so I slightly changed the import hierarchy in daal4py/onedal4py/sklearnex namespaces.
But @icfaust raised an interesting question.
If we look at this issue more broadly, should we generally retain the support_usm_ndarray functionality for daal4py in general? 

Which estimator with daal4py backend do we use for GPU offloading? correct me if nothing other than Kmeans. Kmeans is in the active phase of moving out preview #1770, which should happen in the near future, perhaps already before the next release. In that case, does it make sense to support daal4py for usm_ndarray interop at all? if not, then it makes sense to make changes in my PR and transfer the necessary things for device offloading to onedal4py backend and reuse it in sklearnex.
Removing this from daal4py._device_offloading and adding it to onedal4py._device_offloading also solves the circular imports problem.

After the discussion we will decide to integrate is as is, or update it for onedal4py backend only.

samir-nasibli commented:
Dear reviewers,

I am returning this PR to development.
I will remove all usm_ndarray support from daal4py. All device_offload primitives will be located in the onedal4py module.
Since Kmeans only uses daal4py's GPU offloading, it is only expected that these tests will fail until Kmeans with onedal4py is integrated #1770

samir-nasibli commented on daal4py/sklearn/_config.py:
```diff
@@ -0,0 +1,72 @@
+# ==============================================================================
```
outdated.

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
```
/tests/daal4py/sycl/gradient_boosted_regression.py
FAILED	gradient_boosted_regression.py	with errno	-9
```
Internal CI is green.
Only fail is not related with the changes from this PR.

samir-nasibli commented:
/intelci: run

ethanglaser commented on sklearnex/linear_model/ridge.py:
```diff
@@ -15,3 +15,10 @@
+# TODO:
+# implement GPU offloading via onedal4py backend.
+Ridge.fit = support_usm_ndarray(queue_param=False)(Ridge.fit)
+Ridge.predict = support_usm_ndarray(queue_param=False)(Ridge.predict)
```
General comment for several files - why is it necessary to decorate these functions?

samir-nasibli replied:
For dpnp/dpctl, aka usm_ndarrays interop.

ethanglaser commented on sklearnex/_config.py:
```diff
@@ -46,7 +34,7 @@ def get_config():
     """
     sklearn = skl_get_config()
-    sklearnex = _get_sklearnex_threadlocal_config().copy()
+    sklearnex = _get_onedal_threadlocal_config().copy()
```
would calling _get_config() make more sense?

samir-nasibli replied:
good catch. I think it make sense

samir-nasibli replied:
thank you for the proposal! updated onedal's _get_config a bit for it

ethanglaser commented on sklearnex/dispatcher.py:
```diff
@@ -143,6 +146,10 @@ def get_patch_map_core(preview=False):
         from .linear_model import LinearRegression as LinearRegression_sklearnex
         from .linear_model import LogisticRegression as LogisticRegression_sklearnex
+        from .manifold import TSNE as TSNE_sklearnex
+        from .metrics import pairwise_distances as pairwise_distances_sklearnex
```
I guess elasticnet and ridge would be needed here as well? I am a bit confused on if this is necessary to add to patching but if so then probably need all? @Alexsandruss any idea why these would not be in here?

samir-nasibli replied:
In my understanding, if we wrap this functionality for functional support with dpnp/dpctl, then it matters to add it to the patching. Since now the possibility of wrapping appeared only in sklearnex level, it makes sense. If there are any objections or arguments against, please share

Alexsandruss replied:
All decorated sklearnex estimators should be added in patching map, otherwise unchanged daal4py estimators will be used in it.

samir-nasibli replied:
So this confirms that it is need to update the patching map.

ethanglaser replied:
Yes so wouldn't it make sense to also add ridge and elasticnet to the patching if we are adding tsne?

samir-nasibli replied:
Hey, sorry I misunderstood your original comment. Very good catch. I thought I have already added them. 
Actually we **should** add them to the map. I will update map patching. 

I would like also understand why this tests passed without it. It is weird little bit for me that dpnp/dpctl tests passed… I will check them all once again and report. Potentially there could be an issue.

samir-nasibli replied:
So there was not any `dpnp`/`dpctl` testing with GPU for `Ridge`, `ElasticNet` and `Lasso`. That is why old patching worked fine without `usm_ndarray_support` decoration. Will cover this asap.

samir-nasibli replied:
Pleas see: https://github.com/intel/scikit-learn-intelex/pull/1891

ethanglaser reviewed (commented):
some initial comments below

icfaust commented on sklearnex/linear_model/coordinate_descent.py:
```diff
@@ -15,3 +15,12 @@
 # ===============================================================================
 
 from daal4py.sklearn.linear_model import ElasticNet, Lasso
+from onedal._device_offload import support_usm_ndarray
+
+ElasticNet.fit = support_usm_ndarray(queue_param=False)(ElasticNet.fit)
+ElasticNet.predict = support_usm_ndarray(queue_param=False)(ElasticNet.predict)
+ElasticNet.score = support_usm_ndarray(queue_param=False)(ElasticNet.score)
+
+Lasso.fit = support_usm_ndarray(queue_param=False)(Lasso.fit)
+Lasso.predict = support_usm_ndarray(queue_param=False)(Lasso.predict)
+Lasso.score = support_usm_ndarray(queue_param=False)(Lasso.score)
```
```suggestion
from onedal._device_offload import support_usm_ndarray

class ElasticNet(daal4py_ElasticNet):
    fit = support_usm_ndarray(queue_param=False)(daal4py_ElasticNet.fit)
    predict = support_usm_ndarray(queue_param=False)(daal4py_ElasticNet.predict)
    score = support_usm_ndarray(queue_param=False)(daal4py_ElasticNet.score)

class Lasso(daal4py_Lasso):
    fit = support_usm_ndarray(queue_param=False)(daal4py_Lasso.fit)
    predict = support_usm_ndarray(queue_param=False)(daal4py_Lasso.predict)
    score = support_usm_ndarray(queue_param=False)(daal4py_Lasso.score)

```

samir-nasibli replied:
In this case I also should bring docstrings from daal4py, am I right?

samir-nasibli replied:
But I think this potentially could be an issues that here used only daal4py imported estimator. However, I should check that the methods here are indeed overridden

samir-nasibli replied:
This kind of decorations works fine.
It doesn't change current daal4py, sklearnex dispatch mapping, except moving `support_usm_ndarray` decoration into sklearnex level. See also: [support_usm_ndarray](https://github.com/intel/scikit-learn-intelex/pull/1832/files#r1649469866)

icfaust commented on sklearnex/linear_model/logistic_regression.py:
```diff
@@ -21,18 +21,6 @@
 from daal4py.sklearn.linear_model.logistic_path import (
     LogisticRegression as LogisticRegression_daal4py,
 )
-from daal4py.sklearn.linear_model.logistic_path import daal4py_fit, daal4py_predict
```
Separate PR for this change? Seems unrelated.

samir-nasibli replied:
Sure. This is totally unrelated :)
I actually thought about this. Now I will definitely do a separate PR.
Good catch thank you!

samir-nasibli replied:
Done. Please check #1877

icfaust commented on sklearnex/manifold/t_sne.py:
```diff
@@ -15,3 +15,7 @@
 # ===============================================================================
 
 from daal4py.sklearn.manifold import TSNE
+from onedal._device_offload import support_usm_ndarray
+
+TSNE.fit = support_usm_ndarray(queue_param=False)(TSNE.fit)
+TSNE.fit_transform = support_usm_ndarray(queue_param=False)(TSNE.fit_transform)
```
```suggestion
from onedal._device_offload import support_usm_ndarray

class TSNE(daal4py_TSNE):
    fit = support_usm_ndarray(queue_param=False)(daal4py_TSNE.fit)
    fit_transform = support_usm_ndarray(queue_param=False)(daal4py_TSNE.fit_transform)

```

samir-nasibli replied:
See please discussion: https://github.com/intel/scikit-learn-intelex/pull/1832/files#r1649469866

icfaust reviewed (commented):
Here is a follow-up review.

samir-nasibli commented on onedal/_device_offload.py:
```diff
@@ -14,28 +14,137 @@
+            self._filter_string = filter_string
+            self.is_cpu = "cpu" in filter_string
+            self.is_gpu = "gpu" in filter_string
+            # TODO: check for possibility of fp64 support
```
It is necessary when we are trying to offload GPU via config_context, when dpctl is not in env.
I will cover this with JIRA ticket. I was not thinking to change a lot in the offloading flow. Make sense to work on it seperatly.

samir-nasibli replied:
Covered with ticket. Will remove TODO comment from the source after last CI check completed.

samir-nasibli replied:
> DummySyclQueue and why its necessary to set has_aspect_fp64 to false. Wouldn't a warning or throwing an error be better? I know this a copy of sklearnex/_device_offload,py, but it may be a good time to address this.

I don't think that this is a good time to address. This should be investigated and adressed separately. I don't want it be a blocker. This is just copy of original impl. TODO comment already covered by related ticket, that should be prioritized and adressed.

samir-nasibli commented on onedal/_device_offload.py:
```diff
@@ -59,24 +168,31 @@ def _run_on_device(func, obj=None, *args, **kwargs):
 
-def support_usm_ndarray(freefunc=False):
+if dpnp_available:
+    # TODO:
```
Yeah this will be addressed on #1813

samir-nasibli commented on onedal/common/_policy.py:
```diff
@@ -46,33 +42,19 @@ def _get_queue(*data):
-            self._d4p_interop = _Daal4PyContextReset()
-            if "sklearnex" in sys.modules:
-                from sklearnex._device_offload import DummySyclQueue
+            from onedal._device_offload import DummySyclQueue
```
We don't have any init file for policies imports and it seems to me that this import should be isolated, since in the policy module I do not want to add unnecessary things for import.

If there are any suggestions for the refactoring, we can create tickets for that.

I have already shared perf validations, no regressions here.

Alexsandruss replied:
I don't see any reason why it should not be directly imported after line 50 `if _is_dpc_backend:`

samir-nasibli commented on onedal/_config.py:
```diff
@@ -0,0 +1,46 @@
+# limitations under the License.
+# ==============================================================================
+
+"""Tools to expose some sklearnex's config settings to onedal4py level."""
```
I don't see any problems with the names matching and sklearnex level uses and extends necessary for sklearnex.
We have _device_offload modules in both packages, for example.

If there are any more specific concerns about this, I would be glad to know. If there is a better description, please suggest it.

samir-nasibli replied:
no objections. marking as resolved.

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
@ethanglaser @icfaust thank you for the review!

I have addressed almost to all comments.
I am pleased to invite you to the second round of reviews. Please mark comments as resolved if any.

samir-nasibli commented:
/intelci: run

Alexsandruss commented:
/azp run CI

Alexsandruss commented on daal4py/sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -736,7 +732,7 @@ def dual_gap_(self, value):
         self._gap = None
 
-    score = support_usm_ndarray()(ElasticNet_original.score)
+    score = ElasticNet_original.score
```
Python class inheritance should work fine without this line
```suggestion
```

samir-nasibli replied:
Indeed python class inheritance should work for this. Good catch. These changes are the result of blind removal of the decorator. In reality, there is no need for this assignments. Thank you

Alexsandruss commented on daal4py/sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -853,7 +847,7 @@ def dual_gap_(self, value):
         self._gap = None
 
-    score = support_usm_ndarray()(Lasso_original.score)
+    score = Lasso_original.score
```
```suggestion
```

Alexsandruss commented on daal4py/sklearn/linear_model/_ridge.py:
```diff
@@ -314,15 +313,13 @@ def __init__(
         return _predict_ridge(self, X)
 
-    score = support_usm_ndarray()(Ridge_original.score)
+    score = Ridge_original.score
```
```suggestion
```

Alexsandruss commented on sklearnex/linear_model/coordinate_descent.py:
```diff
@@ -15,3 +15,12 @@
 from daal4py.sklearn.linear_model import ElasticNet, Lasso
+from onedal._device_offload import support_usm_ndarray
+
+ElasticNet.fit = support_usm_ndarray(queue_param=False)(ElasticNet.fit)
+ElasticNet.predict = support_usm_ndarray(queue_param=False)(ElasticNet.predict)
+ElasticNet.score = support_usm_ndarray(queue_param=False)(ElasticNet.score)
```
While I don't think this decoration way will cause any problem, it creates situation where two classes with one module path and name exist in daal4py and sklearnex.

samir-nasibli replied:
In fact, this situation existed before these changes. With this changes, here only added decoration usm ndarray support

icfaust replied:
@Alexsandruss Doing this way will make errors to our users more difficult to disentangle. The trace of an error will lead through the wrapping which is not related to the module file in which it is defined. If they see any error with the wrapping whatsoever, they will have to really work to find how and where it is included.

samir-nasibli replied:
@icfaust personally I agree with this your proposal. It make sense in case of debug.
Will address and create exact sklearnex estimators with daal4py inheritance.

Alexsandruss commented on sklearnex/linear_model/coordinate_descent.py:
```diff
@@ -15,3 +15,12 @@
 from daal4py.sklearn.linear_model import ElasticNet, Lasso
+from onedal._device_offload import support_usm_ndarray
+
+ElasticNet.fit = support_usm_ndarray(queue_param=False)(ElasticNet.fit)
+ElasticNet.predict = support_usm_ndarray(queue_param=False)(ElasticNet.predict)
+ElasticNet.score = support_usm_ndarray(queue_param=False)(ElasticNet.score)
```
JFYI: __doc__ is saved during this decoration.

samir-nasibli replied:
Yes, that is right. Therefore unnecessary doc assignments are not added.

Alexsandruss commented on sklearnex/dispatcher.py:
```diff
@@ -143,6 +146,10 @@ def get_patch_map_core(preview=False):
         )
         from .linear_model import LinearRegression as LinearRegression_sklearnex
         from .linear_model import LogisticRegression as LogisticRegression_sklearnex
```
Absence of decorated Lasso, Ridge, ElasticNet in sklearnex patch_map creates uncertainty: is decorated sklearnex class used or undecorated daal4py one?

samir-nasibli replied:
See https://github.com/intel/scikit-learn-intelex/pull/1832#discussion_r1648578243 . Thank you!

samir-nasibli commented on sklearnex/dispatcher.py:
Kmeans fix on #1875

samir-nasibli commented:
/azp run CI

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
@ethanglaser @icfaust @Alexsandruss thank you for the review!
Dear reviewers I have addressed all your comments. Please mark comments as resolved if any.

ethanglaser commented on sklearnex/dispatcher.py:
```diff
@@ -201,6 +237,54 @@ def get_patch_map_core(preview=False):
+        ]
+
+        # metrics
+        mapping.pop("distances")
+        mapping.pop("roc_auc_score")
+        mapping["distances"] = [
+            [
+                (metrics_module, "pairwise_distances", pairwise_distances_sklearnex),
+                None,
+            ]
+        ]
+        mapping["pairwise_distances"] = mapping["distances"]
```
Why is "distances" key needed?

samir-nasibli replied:
I am not sure why is it needed. But it existed in daal4py mapping.
For this kind of cases I just made consistent changes: see https://github.com/intel/scikit-learn-intelex/blob/324c9ccf928d439c64a2759f60c5bcc782dd5a81/sklearnex/dispatcher.py#L238C9-L238C59

ethanglaser commented on sklearnex/_device_offload.py:
```diff
@@ -14,16 +14,20 @@
 
-import logging
-import sys
 from collections.abc import Iterable
```
```suggestion
```

ethanglaser commented:
No major objections from my side at this point, looking good, seems like the queue/device offload will functionally remain the same. One remaining question I have is what separates functionality defined in onedal/_device_offload.py vs. sklearnex/_device_offload.py at this point? Seems like maybe these could all be moved to onedal4py. But if there is functionality in sklearnex for a reason then I guess not.

samir-nasibli commented:
> No major objections from my side at this point, looking good, seems like the queue/device offload will functionally remain the same. One remaining question I have is what separates functionality defined in onedal/_device_offload.py vs. sklearnex/_device_offload.py at this point? Seems like maybe these could all be moved to onedal4py. But if there is functionality in sklearnex for a reason then I guess not.

Thank you!
sklearnex specific functional I left in sklearnex._device_offload module. I think it should be ok now.

ethanglaser commented on sklearnex/_device_offload.py:
```diff
@@ -38,117 +42,6 @@
     dpnp_available = False
```
can just add `import dpnp` to line 27 conditional block

samir-nasibli replied:
dpnp/dcptl import I would like to refactor in general in the project, but for the current changes it is really outdated. Good catch will change.

samir-nasibli commented:
/intelci: run

ethanglaser reviewed (approved):
LGTM, good work on this and good overall review process, let's wait for internal CI to make sure

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
Last CI run: http://intel-ci.intel.com/ef3451eb-4a87-f13b-98de-a4bf010d0e2e

samir-nasibli commented:
/intelci: run

samir-nasibli commented:
CI looks good. Fails are not related with changes on current PR. Merging it
Many thanks to the reviewers for the excellent work done.
